<a href="https://colab.research.google.com/github/ohimjosh/CSC-44800/blob/main/AS03/a3_sample_notebook.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

\(\^Be sure to update this button to point to your notebook instead of the sample notebook\)

In [144]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

## Part 1. Loading the dataset

In [145]:
# Using pandas load the dataset (load remotely, not locally)
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

data = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
data.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [146]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


## Part 2. Splitting the dataset

In [147]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = data[["Temperature °C", "Mols KCL"]]
y = data["Size nm^3"]
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.9, test_size=0.1)

In [148]:
X

Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]

In [149]:
y

0      6.244743e+05
1      5.779610e+05
2      6.196847e+05
3      1.460449e+06
4      4.325726e+04
           ...     
995    1.545661e+06
996    6.737041e+05
997    3.477543e+05
998    8.684794e+05
999    8.476413e+05
Name: Size nm^3, Length: 1000, dtype: float64

In [150]:
X_train

Temperature °C  Mols KCL
601             844       207
902             752       701
141             423       820
698             356       497
547             455       617
..              ...       ...
722             810       436
378             444        15
226             334         4
405             911       558
634              51       233

[900 rows x 2 columns]

In [151]:
X_test

Temperature °C  Mols KCL
255             513        85
211             815        53
679             770        12
96               86       443
513             277       497
..              ...       ...
996             327       982
440             608       465
812             733       628
934             170       140
364             474       562

[100 rows x 2 columns]

In [152]:
y_train


601    3.607683e+05
902    1.077368e+06
141    7.180074e+05
698    3.651934e+05
547    5.778068e+05
           ...     
722    7.214713e+05
378    1.865443e+04
226    6.680457e+03
405    1.036504e+06
634    2.592911e+04
Name: Size nm^3, Length: 900, dtype: float64

In [153]:
y_test

255     93572.42857
211     96250.25714
679     27724.11429
96      82835.11429
513    285719.40000
           ...     
996    673704.11430
440    578913.85710
812    940712.11430
934     50200.00000
364    547488.11430
Name: Size nm^3, Length: 100, dtype: float64

## Part 3. Perform a Linear Regression

In [154]:
# Use sklearn to train a model on the training set
reg = LinearRegression().fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model

reg.predict(np.array([[300,350]]))

/home/ohimjosh/programming/CSC-44800/AS03/venv/lib/python3.8/site-packages/sklearn/base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([205671.39930347])

In [155]:
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
reg.score(X_train, y_train)

0.8581981292498426

In [160]:
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
reg.coef_

array([ 876.43826102, 1025.93060778])

In [161]:
reg.intercept_

-416335.79172673094

h(x) : y = mx + b

h(x) : $Θ_{0}$ + $Θ_{1}x$

h(x) : 876.43826102 * $x_{1}$ + 1025.93060778 * $x_{2}$ - 416335.79172673094


Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [157]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

## Part 5. Using Polynomial Regression

In [158]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2

# Report on the metrics and output the resultant equation as you did in Part 3.